lib/engine/operator.rb

```diff
@@ -60,5 +61,37 @@ def unplaced_tokens
     def placed_tokens
       @tokens.select(&:city)
     end
+
+    def current_book_value
+      trains = @trains.sum(&:price)
+      tokens = if @token_book_value_override
+                 @tokens.size * @token_book_value_override
+               else
+                 @tokens.filter(&:used).sum(&:price)
+               end
+      loans = @loans.sum(&:amount)
+      trains + tokens + @cash - loans
+    end
+
+    def rusted_book_value(depot)
+      rust_value = 0
+
+      rust_schedule, obsolete_schedule = depot.rust_obsolete_schedule
+      unless depot.upcoming.empty?
+        depot.upcoming.group_by(&:name).map do |name, _|
+          rusted_trains = rust_schedule[name] || []
+          obsolete_trains = obsolete_schedule[name] || []
+
+          trains = rusted_trains + obsolete_trains
+
+          unless trains.empty?
+            rust_value = @trains.filter { |t| trains.include?(t.name) }.sum(&:price)
```

Comment:
same

Reply:
in this case you can just do  @trains.sum { |t| trains.include?(t.name) ? t.price : 0 } so you only do one iteration